PyRoute/Inputs/StarlineTransformer.py

```diff
@@ -0,0 +1,129 @@
+
+    def transform(self, tree):
+        parsed = {'ix': None, 'ex': None, 'cx': None, 'residual': ''}
+
+        for kid in tree.children:
+            if isinstance(kid, Token):
+                parsed['trade'].append(kid.value)
+                continue
+
+            dataval = str(kid.data)
+            if 'starname' == dataval:
+                rawval = kid.children[0].value
+                result = self.starname(rawval)
+                parsed['name'] = result[0]
+                parsed['uwp'] = result[1]
+            elif 'trade' == dataval:
+                parsed['trade'] = self.trade(kid)
+            elif 'extensions' == dataval:
+                parsed['ix'], parsed['ex'], parsed['cx'] = self.extensions(kid)
+            elif 'world_alg' == dataval:
+                parsed['pbg'], parsed['worlds'], parsed['allegiance'] = self.world_alg(kid)
+            else:
+                rawval = kid.children[0].value
+                parsed[dataval] = rawval
+
+        self.square_up_trail_dash_in_trade_codes(parsed)
+        self.square_up_nobles_baze_and_zone(parsed)
+        self.square_up_worlds_alg_and_residual(parsed)
+
+        for line in self.strip_list:
+            parsed[line] = parsed[line].strip()
+
+        no_extensions = parsed['ix'] is None and parsed['ex'] is None and parsed['cx'] is None
+        ex_ix = parsed['ix'] if parsed['ix'] is not None else ' '
+        ex_ix = self.boil_down_double_spaces(ex_ix)
+        extensions = (ex_ix) + ' ' + \
+                     (parsed['ex'] if parsed['ex'] is not None else ' ') + ' ' + \
+                     (parsed['cx'] if parsed['cx'] is not None else ' ')
+        spacer = ' ' if no_extensions else None
+        if no_extensions:
+            extensions = ''
+        # Currently aiming to drop-in replace the starline regex output
+        data = [parsed['position'], parsed['name'], parsed['uwp'], parsed['trade'], extensions, parsed['ix'], parsed['ex'], parsed['cx'], spacer, spacer, spacer, parsed['nobles'], parsed['base'], parsed['zone'], parsed['pbg'], parsed['worlds'], parsed['allegiance'], parsed['residual']]
+
+        return data
+
+    def square_up_worlds_alg_and_residual(self, parsed):
+        if 3 < len(parsed['pbg']):
+            return
+        if ' ' != parsed['worlds']:
+            return
+        if not parsed['allegiance'].strip().isdigit():
+            return
+        if 2 > len(parsed['residual'].strip()):
+            return
+        parsed['worlds'] = parsed['allegiance']
+        parsed['allegiance'] = parsed['residual']
+        parsed['residual'] = ''
+
+    def square_up_nobles_baze_and_zone(self, parsed):
+        zonelen = len(parsed['zone'])
+        if 4 > zonelen or '' != parsed['nobles'].strip():
+            return
+        parsed['nobles'] = parsed['base']
+        parsed['base'] = parsed['zone']
+        parsed['zone'] = ' '
+
+    @staticmethod
+    def boil_down_double_spaces(dubbel):
```

Comment:
This could be replaced with `" ".join(dubbel.split())` 
The `split()` breaks the string on spaces (removing the extras) and the join put the elements back together with one space between them. Which is, according to my reading, what you are attempting to do here.

Reply:
Didn't realise that.  Will give it a go.  Thanks

Reply:
Like so?

Reply:
Yes. This seems to be passing the tests.